packages/plugin-ext/doc/how-to-add-new-custom-plugin-api.md

```diff
@@ -0,0 +1,281 @@
+export const PLUGIN_RPC_CONTEXT = {
+    FOO_MAIN: createProxyIdentifier<FooMain>('FooMain')
+};
+
+// Main application will obtain a proxy using these IDs, plugin host will register an implementation for it.
+export const MAIN_RPC_CONTEXT = {
+    FOO_EXT: createProxyIdentifier<FooExt>('FooExt')
+};
+```
+
+On the plugin host side we can register our implementation and retrieve the proxy as part of our `createAPIFactory` implementation:
+
+Example `foo-ext.ts`:
+
+```typescript
+export class FooExtImpl implements FooExt {
+    // Main application RCP counterpart
+    private proxy: FooMain;
+
+    constructor(rpc: RPCProtocol) {
+        rpc.set(MAIN_RPC_CONTEXT.FOO_EXT, this); // register ourselves
+        this.proxy = rpc.getProxy(PLUGIN_RPC_CONTEXT.FOO_MAIN); // retrieve proxy
+    }
+
+    getFooImpl(): Promise<Foo> {
+        return this.proxy.$getFooImpl();
+    }
+}
+```
+
+On the main side we need to implement the counterpart of the ExtPluginApiProvider, the `MainPluginApiProvider`, and expose it in a browser frontend module:
+
+Example `foo-main.ts`:
+
+```typescript
+@injectable()
+export class FooMainImpl implements FooMain {
+    @inject(MessageService) protected messageService: MessageService;
+    protected proxy: FooExt;
+
+    init(rpc: RPCProtocol) {
+        // We would use this if we had a need to call back into the plugin-host/plugin
+        this.proxy = rpc.getProxy(MAIN_RPC_CONTEXT.FOO_EXT);
+    }
+
+    async $getFooImpl(): Promise<Foo> {
+        this.messageService.info('We were called from the plugin-host at the behest of the plugin.');
+        return new Foo();
+    }
+}
+
+@injectable()
+export class FooMainPluginApiProvider implements MainPluginApiProvider {
+    @inject(MessageService) protected messageService: MessageService;
+
+    initialize(rpc: RPCProtocol, container: interfaces.Container): void {
+        this.messageService.info('Initialize RPC communication for FooMain!');
+        // create a new FooMainImpl as it is not bound as singleton
+        const fooMainImpl = container.get(FooMainImpl);
+        fooMainImpl.init(rpc);
+        rpc.set(PLUGIN_RPC_CONTEXT.FOO_MAIN, fooMainImpl);
+    }
+}
+
+export default new ContainerModule(bind => {
+    bind(FooMainImpl).toSelf();
+    bind(MainPluginApiProvider).to(FooMainPluginApiProvider).inSingletonScope();
+});
+```
+
+In this example, we can already see the big advantage of going to the main application side as we have full access to our Theia services.
+
+## Usage in a plugin
+
+When using the API in a plugin the user can simply use the API as follows:
+
+```typescript
+import * as foo from '@bar/foo';
+
+foo.fooBar.getFoo();
+```
+
+## Packaging
+
+When bundling our application with the generated `gen-webpack.node.config.js` we need to make sure that our initialization function is bundled as a `commonjs2` library so it can be dynamically loaded.
```

Comment:
```suggestion
When bundling our application with the generated `gen-webpack.node.config.js` we need to make sure that our initialization function is bundled as a `commonjs2` library so it can be dynamically loaded. Adjust `webpack.config.js` accordingly:
```